### Term Project  
2022105661 강민성

#### 주제 : PUBG(배틀그라운드)에서 플레이어 성과 분석

##### (1) 주제 선정 이유
평소 PUBG(배틀그라운드)를 자주 플레이하고 최근에 PUBG에 대해 조사를 진행하며 게임 개발에 대한 관심도 또한 증가하게 되었습니다.<br>
그래서 이번 Term Project를 PUBG 관련 주제로 진행하고자 해서 추가적으로 웹 서칭을 실시했습니다.<br>
서칭해본 결과, 특정 PUBG 플레이어의 평균 등수, 평균 킬 수, 평균 생존 시간 등을 분석해 보고 싶어서 주제를 <br>플레이어 성과 분석으로 정하게 되었습니다.


##### (2) 가설 정의
- 킬 수와 등수 간의 상관관계 :  
킬 수가 많은 플레이어일수록 등수가 높을 가능성이 크다.
그러나, 킬 수가 많아도 생존 시간이 짧을 경우, 등수가 낮을 수 있다.

- 생존 시간과 등수 간의 상관관계 :  
생존 시간이 긴 플레이어일수록 등수가 높을 가능성이 크다.
그러나, 생존 시간이 길어도 킬 수가 낮다면 등수가 낮을 수도 있다.

- 킬 수와 생존 시간 간의 상관관계 :  
킬 수가 많은 플레이어는 보통 교전이 잦기 때문에 생존 시간이 짧을 가능성이 있다.
반면, 킬 수가 적은 플레이어는 교전을 피하고 생존 시간이 길어질 가능성이 있다.


##### (3) 인터넷을 통한 데이터 획득

킬, 생존시간, 순위의 상관관계를 확인하기 위헤 플레이어 ID를 통해 해당 플레이어의 최근 n번의 매치에서 킬, 생존시간, 순위 데이터를 수집했습니다.<br>
PUBG developer에 API를 요청하여 API key를 확보했습니다. 그리고 PUBG API에서 플레이어 ID -> 매치 ID 목록 -> 매치 데이터의 순서로 데이터를 수집하였습니다.<br>
 다음으로, 각 매치에서 킬 수, 생존 시간, 등수를 추출하여 CSV 파일로 저장하였습니다. <br>

PUBG developer에 API를 요청하여 API key를 확보했습니다. 하지만, 해당 api 키를 활용해서 `플레이어 ID를 통해 match Id를 받고, 매치 데이터를 수집, 가공하는 과정은<br>
제 수준에 하기에는 어려움이 존재하는 작업이라고 판단 후, 생성형 ai에게 해당 작업을 완료하는 코드를 의뢰했습니다.<br>
의뢰 내용은 모든 플레이어의 데이터를 확인하는 것은 불가능에 가깝기에 유명 스트리머 4명의 최근 20게임 매치 결과를 csv파일에 저장하는 것 입니다.<br>
(해당 코드는 match_data_get.py로 저장했습니다.)

player_stats.csv에 저장된 데이터를 불러왔습니다.

In [1]:
import pandas as pd

csv_file = "player_stats.csv"

df = pd.read_csv(csv_file)

print(df)

          name  kills  timeSurvived  rank name_kor
0   Zucchini__      2          1461     2      주키니
1   Zucchini__      0          1278     9      주키니
2   Zucchini__      0           811    14      주키니
3   Zucchini__      0           236    29      주키니
4   Zucchini__      0           542    17      주키니
..         ...    ...           ...   ...      ...
75     kimblue      0           243    28      김블루
76     kimblue      0          1073    16      김블루
77     kimblue      0           579    25      김블루
78     kimblue      6          1548     1      김블루
79     kimblue      0           165    48      김블루

[80 rows x 5 columns]


##### (4) 분석을 위한 데이터의 가공

분석을 위해서 굳이 name_kor은 필요가 없다고 생각해서 name, kills, timeSurvived, rank <br>
이렇게 필요한 4가지만 사용하였습니다.<br>
그리고 csv에 저장되어있는 숫자가 문자열인지 정수형인지 잘 몰라서 해당 숫자들을 전부 정수형으로 변경했습니다.<br>
이러한 과정을 통해 데이터가 평균을 구하기 위한 준비를 완료했습니다.

In [2]:
import pandas as pd

df = pd.read_csv("player_stats.csv")

name = list(df["name"])
kills = list(df["kills"])
timeSurvived = list(df["timeSurvived"])
rank = list(df["rank"])

new_kills = []
new_timeSurvived = []
new_rank = []

for i in range(len(df)):
    new_kills.append(int(kills[i]))
    new_timeSurvived.append(int(timeSurvived[i]))
    new_rank.append(int(rank[i]))

new_df = pd.DataFrame({
    "name": name,
    "kills": new_kills,
    "timeSurvived": new_timeSurvived,
    "rank": new_rank
})

print(df)

          name  kills  timeSurvived  rank name_kor
0   Zucchini__      2          1461     2      주키니
1   Zucchini__      0          1278     9      주키니
2   Zucchini__      0           811    14      주키니
3   Zucchini__      0           236    29      주키니
4   Zucchini__      0           542    17      주키니
..         ...    ...           ...   ...      ...
75     kimblue      0           243    28      김블루
76     kimblue      0          1073    16      김블루
77     kimblue      0           579    25      김블루
78     kimblue      6          1548     1      김블루
79     kimblue      0           165    48      김블루

[80 rows x 5 columns]


##### (5) 분석 결과 도출

가공된 데이터를 기반으로 분석을 진행했습니다.<br>
1. 평균 킬 수 계산<br>
2. 평균 생존 시간 계산<br>
3. 평균 순위 계산<br>

In [1]:
import pandas as pd

class PlayerStats:
    def __init__(self, name):
        self.name = name
        self.total_kills = 0
        self.total_timeSurvived = 0
        self.total_rank = 0
        self.match_count = 0

    def add_match(self, kills, timeSurvived, rank):
        self.total_kills += int(kills)
        self.total_timeSurvived += int(timeSurvived)
        self.total_rank += int(rank)
        self.match_count += 1

    def get_averages(self):
        if self.match_count == 0:
            return 0, 0, 0
        avg_kills = self.total_kills / self.match_count
        avg_timeSurvived = self.total_timeSurvived / self.match_count
        avg_rank = self.total_rank / self.match_count
        return avg_kills, avg_timeSurvived, avg_rank

def analyze_player_stats(file_path):
    df = pd.read_csv(file_path)

    players = {}

    for _, row in df.iterrows():
        name = row["name"]
        kills = row["kills"]
        timeSurvived = row["timeSurvived"]
        rank = row["rank"]

        if name not in players:
            players[name] = PlayerStats(name)

        players[name].add_match(kills, timeSurvived, rank)

    for name, stats in players.items():
        avg_kills, avg_timeSurvived, avg_rank = stats.get_averages()
        print(f"{name}: 평균 킬 {round(avg_kills, 2)}, 평균 생존시간 {round(avg_timeSurvived, 2)}초, 평균 순위 {round(avg_rank, 2)}")

analyze_player_stats("player_stats.csv")

Zucchini__: 평균 킬 1.2, 평균 생존시간 1123.25초, 평균 순위 13.7
obbayasalido: 평균 킬 3.1, 평균 생존시간 1125.85초, 평균 순위 25.55
rnsflaqh: 평균 킬 3.65, 평균 생존시간 1166.85초, 평균 순위 17.4
kimblue: 평균 킬 3.4, 평균 생존시간 761.3초, 평균 순위 32.1


##### (6) 결론


총 4명의 유명 스트리머를 분석했습니다.<br>
1. 군림보 - player_name = "rnsflaqh"<br>
2. 주키니 - player_name = "Zucchini__"<br>
3. 미라클 - player_name = "obbayasalido"<br>
4. 김블루 - player_name = "kimblue"<br>

| 스트리머       | 킬 수 평균 | 생존 시간 평균(초) | 평균 순위 |
|----------------|------------|---------------------|------------|
| 군림보         | 3.65        | 1166.85              | 17.4       |
| 주키니         | 1.2        | 1123.25              | 13.7       |
| 미라클         | 3.1        | 1125.85               | 25.55       |
| 김블루     | 3.4        | 761.3               | 32.1        |

##### 1. 킬 수와 순위의 상관관계
당연히 킬 수가 많을수록 순위가 높을 가능성이 크지만,<br>
생존시간이 짧다면 순위가 낮을수도 있습니다.<br>
'군림보'를 보면 평균 킬 수가 가장 높고 순위도 상위권입니다.<br>
하지만, '김블루'와 '미라클'은 높은 킬 수에도 낮은 평균 순위를 갖고 있습니다.<br>
따라서, **킬 수가 높다고 무조건 높은 순위를 보장하는건 아니다** 라는 결론을 도출했습니다.

##### 2. 생존 시간과 순위의 상관관계
가설은 "생존 시간이 길수록 순위가 높을 것이다."입니다.<br>
'주키니'는 높은 편의 생존시간과 가장 높은 순위를 기록해서 가설과 일치합니다.<br>
'군림보','주키니','미라클'은 비슷한 생존 시간과 어느 정도 높은 순위를 기록합니다.<br>
그리고 '김블루'는 가장 낮은 생존시간과 가자 낮은 순위로 가설의 대우 사례입니다.<br>
따라서 **생존시간이 길면 순위가 높을 가능성이 크다** 라는 결론을 도출했습니다.


#### 3. 킬 수와 생존 시간의 상관관계
가설은 '킬 수가 많으면 잦은 교전으로 인한 것이기 때문에 생존시간은 짧다'입니다.<br>
'김블루'는 많은 킬과 짧은 생존시간으로 가설과 일치합니다.<br>
'군림보'와 '미라클'은 많은 킬과 긴 생존시간을 보여줍니다.<br>
'주키니'는 적인 킬과 긴 생존 시간을 보여줍니다.<br>
따라서 일부 플레이어는 일치하지만 일부는 아니기 때문에 얻은 결론은<br>
**일부는 가설과 일치하지만, 많은 킬을 하면서도 긴 생존시간을 기록하는 것도 가능하다** 입니다.

#### 총 결론
**킬 수와 등수는 생존 시간과 함께 고려해야 한다**는 점이 확인되었습니다.<br>
결론적으로, PUBG에서 플레이어의 성능을 평가할 때는<br>
 **<단일 지표가 아닌, 킬 수, 생존 시간, 순위를 종합적으로 해석>**<br>
 하는 것이 타당하다는 것을 알았습니다.

##### (7) 참고문헌

1. PUBG Developer API 공식 사이트트 
   - https://developer.pubg.com/

2. Youtube 스트리머 채널 
   - 주키니 : https://www.youtube.com/@ZucchiniTV
   - 군림보 : https://www.youtube.com/@%EA%B5%B0%EB%A6%BC%EB%B3%B4
   - 미라클 : https://www.youtube.com/@miracle0o0
   - 왓구홍길동 : https://www.youtube.com/@%EC%99%93%EA%B5%AC%ED%99%8D%EA%B8%B8%EB%8F%99

3. Pandas 라이브러리 공식 문서
   - https://pandas.pydata.org/docs/


##### (8) 별첨 : 플레이어 정보를 수집한 원본 데이터의 예시시

스트리머 '군림보'의 매치 데이터 (실제로는 50개 이상이지만 너무 길기 때문에 3개만 나타냈습니다.)<br>
```json
{
  "data": [{
    "type": "player",
    "id": "account.178fea2929b449538f510e51aedbb65b",
    "attributes": {
      "name": "rnsflaqh",
      "banType": "Innocent",
      "shardId": "steam"
    },
    "relationships": {
      "matches": {
        "data": [
          { "type": "match", "id": "1eb60105-c426-45a3-a927-ce6ea661e4a2" },
          { "type": "match", "id": "d7aa6950-fd25-4ad4-9ff5-c944825016d2" },
          { "type": "match", "id": "af3877ed-039d-4dcd-bb1a-3705f63fb38a" },
          ...
        ]
      }
    }
  }]
}


##### (9) 가공된 데이터 예시

분석을 위해 정리한 플레이어 데이터는 수집합 정보<br>
'이름', '킬 수', '생존시간', '순위' 이렇게 필요한 정보만 추출한 것 입니다.<br>
실제로는 80개의 매치 기록이 저장되었지만, 보기 편하게 4개만 뽑아서 작성했습니다.

```python
import pandas as pd

data = [
    {"name": "rnsflaqh", "kills": 3, "time_survived": 774, "rank": 44, "kills_per_minute": 0.23},
    {"name": "Zucchini__", "kills": 2, "time_survived": 1453, "rank": 7, "kills_per_minute": 0.08},
    {"name": "obbayasalido", "kills": 4, "time_survived": 1324, "rank": 13, "kills_per_minute": 0.18},
    {"name": "kimblue", "kills": 4, "time_survived": 1763, "rank": 1, "kills_per_minute": 0.14}
]

df = pd.DataFrame(data)
df.to_csv("player_stats.csv")
print(df)
```

##### (10) 데이터를 분석하기 위해 가공한 코드 원본

API KEY를 통해 획득한 데이터가 저장된 'player_stats.csv'에서 <br>
데이터를 분석해서 가공하는데 사용한 코드 원본이다.

In [4]:
import pandas as pd

df = pd.read_csv("player_stats.csv")

player_name = []

name = list(df["name"])
kill = list(df["kills"])
timeSurvived = list(df["timeSurvived"])
rank = list(df["rank"])


for n in name:
    if n not in player_name:
        player_name.append(n)

sum_kills = {}
sum_timeSurvived = {}
sum_rank = {}
number = {}

for i in range(len(name)):
    n = name[i]
    if n not in sum_kills:
        sum_kills[n] = 0
        sum_timeSurvived[n] = 0
        sum_rank[n] = 0
        number[n] = 0

    sum_kills[n] += int(kill[i])
    sum_timeSurvived[n] += int(timeSurvived[i])
    sum_rank[n] += int(rank[i])
    number[n] += 1

for name in player_name:
    avg_kill = sum_kills[name] / number[name]
    avg_timeSurvived = sum_timeSurvived[name] / number[name]
    avg_rank = sum_rank[name] / number[name]

    print(f"{name}:  평균 킬 {round(avg_kill,2)}, 평균 생존시간 {round(avg_timeSurvived,2)}초, 평균 순위 {round(avg_rank,2)}")

Zucchini__:  평균 킬 1.2, 평균 생존시간 1123.25초, 평균 순위 13.7
obbayasalido:  평균 킬 3.1, 평균 생존시간 1125.85초, 평균 순위 25.55
rnsflaqh:  평균 킬 3.65, 평균 생존시간 1166.85초, 평균 순위 17.4
kimblue:  평균 킬 3.4, 평균 생존시간 761.3초, 평균 순위 32.1
